### PR TITLE
Reuse existing build artefacts where available

### PR DESCRIPTION
**Pull request by shawnlaffan**
This will save substantial development time as long-running build processes can avoid rebuilding dependencies.

### CODE REVIEW

**Review by genio — APPROVED** [review]

File: build.sh [hunk 1]
```diff
@@ -157,7 +157,9 @@ if [ ! -e "$1" ] ; then
 else
   # parameter 1: pkg-list filename
   export PKGLISTNAME=$1
-  PKGLIST=`grep -v -e "^#" -e "^\s*$" "$1" 2>/dev/null`
+#  PKGLIST=`grep -v -e "^#" -e "^\s*$" "$1" 2>/dev/null`
+  PKGLIST=`perl process_build_artefacts.pl "$1" "$2"`
```
Comment (genio) [inline build.sh:161]: Should we be using `$()` instead of backticks at this point?

Comment (genio) [reply build.sh:161]: I know this script is littered with backticks throughout, but maybe it's time to "fix" that? That's likely a different PR, this is more conversation than an ask for changes to this PR.

Comment (shawnlaffan) [reply build.sh:161]: Would be good to modernise it, but I agree that's a separate update.  The whole script could probably also be rewritten in Perl but that's a much larger undertaking.  (I suspect the bash script is a historical artefact from early builds where Perl might not have been available or needed to be avoided).

**Comment by shawnlaffan** [comment]
One limitation of the approach in this script is where a package early in the chain (```pk1```) is rebuilt but the intermediate ones are not, and where the intermediate packages change behaviour based on the version of ```pk1```.  

e.g. given the dependency chain ```pk1 -> pk2 -> pk3 -> pk4```, where ```pk2``` changes properties when compiled against a more recent version of ```pk1```.  If the build script is told to rebuild ```pk1``` as part of a build for ```pk4``` then this script will grab the build artefacts for ```pk2``` and ```pk3```.  

In this case the source file can be updated to add rebuild flags to each package but that gets messy for non-simple dependency chains as people running the script need to know what the dependencies are.  It is probably simpler to add an env var to force a rebuild of all target packages.  

In a perfect world there would be dependency tree for all the packages but that's a separate issue, perhaps on the wishlist end of the spectrum.  (Such a tree could be cribbed from the MSYS2 packaging system).

**Comment by shawnlaffan** [comment]
The script name also needs work as it could be interpreted as a post-processing step.  

Perhaps something like 'filter_source_list_and_reuse_build_artefacts' which is informative but perhaps too long, although it is only for use from the build script.

**Comment by genio** [comment]
I think I'm good with the name as-is. From what I see, this is ready to be merged. Do you have anything else you wanted to fix up before I do so?

**Comment by shawnlaffan** [comment]
Thanks. 
 
I have no more commits for this PR.

**Comment by shawnlaffan** [comment]
Although one potential issue is with non-standard DLL suffixes.  The packaged files in the _out dir do not include the suffix in their names so if a build is run using a suffix of ```__``` and then a second is run using ```xx``` then the second will have a mix of suffixes.  

How much this matters depends on how flexible this system needs to be.  It could just require a doc update to note that all builds in a container should use a consistent suffix for a given bitness.

**Pull request merged**